File: client-kotlin-spring-restclient/client/src/main/kotlin/app/epistola/client/auth/JwtSigner.kt
package app.epistola.client.auth

import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.ECDSASigner
import com.nimbusds.jose.crypto.RSASSASigner
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.SignedJWT
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.nio.file.Files
import java.nio.file.Path
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.interfaces.ECPrivateKey
import java.security.interfaces.RSAPrivateKey
import java.security.spec.PKCS8EncodedKeySpec
import java.time.Duration
import java.time.Instant
import java.util.Base64
import java.util.Date
import java.util.UUID

/**
 * Creates and signs short-lived JWTs for self-signed JWT authentication with Epistola.
 *
 * Each token contains:
 * - `iss`: the consumer ID
 * - `iat`: issued-at timestamp
 * - `exp`: expiry (iat + tokenLifetime)
 * - `jti`: unique nonce (UUID) for replay protection
 *
 * Example:
 * ```
 * val signer = JwtSigner.builder()
 *     .consumerId("invoice-service")
 *     .privateKey(JwtSigner.loadPrivateKey(Path.of("private.pem")))
 *     .build()
 *
 * val restClient = RestClient.builder()
 *     .baseUrl("http://localhost:8080/api")
 *     .requestInterceptor(identity.interceptor())   // User-Agent + X-EP-Node-Id
 *     .requestInterceptor(signer.interceptor())      // Authorization: Bearer
 *     .build()
 * ```
 */
class JwtSigner private constructor(
    private val consumerId: String,
    private val privateKey: PrivateKey,
    private val algorithm: JWSAlgorithm,
    private val tokenLifetime: Duration,
) {
    companion object {
        fun builder(): Builder = Builder()

        /**
         * Loads a private key from a PEM file. Supports RSA and Ed25519 keys
         * in PKCS#8 format (BEGIN PRIVATE KEY).
         */
        fun loadPrivateKey(path: Path): PrivateKey {
            val pem = Files.readString(path)
            return parsePrivateKeyPem(pem)
        }

        /**
         * Parses a PEM-encoded private key string. Supports RSA and Ed25519 keys
         * in PKCS#8 format (BEGIN PRIVATE KEY).
         */
        fun parsePrivateKeyPem(pem: String): PrivateKey {
            val base64 = pem.lines()
                .filter { !it.startsWith("-----") }
                .joinToString("")
            val keyBytes = Base64.getDecoder().decode(base64)
            val keySpec = PKCS8EncodedKeySpec(keyBytes)

            return try {
                KeyFactory.getInstance("RSA").generatePrivate(keySpec)
            } catch (_: Exception) {
                try {
                    KeyFactory.getInstance("EC").generatePrivate(keySpec)
                } catch (_: Exception) {
                    throw IllegalArgumentException(
                        "Failed to parse private key. Supported formats: RSA, EC (P-256) in PKCS#8 PEM format.",
                    )
                }
            }
        }

        private fun detectAlgorithm(key: PrivateKey): JWSAlgorithm = when (key) {
            is RSAPrivateKey -> JWSAlgorithm.RS256
            is ECPrivateKey -> JWSAlgorithm.ES256
            else -> throw IllegalArgumentException(
                "Unsupported key type: ${key::class.simpleName} (algorithm: ${key.algorithm}). " +
                    "Supported: RSA (2048+), EC (P-256)",
            )
        }
    }

    /**
     * Creates a signed JWT token with a fresh `iat`, `exp`, and `jti`.
     */
    fun createToken(): String {
        val now = Instant.now()

        val claims = JWTClaimsSet.Builder()
            .issuer(consumerId)
            .issueTime(Date.from(now))
            .expirationTime(Date.from(now.plus(tokenLifetime)))
            .jwtID(UUID.randomUUID().toString())
            .build()

        val header = JWSHeader.Builder(algorithm).build()
        val jwt = SignedJWT(header, claims)

        val signer = when (algorithm) {
            JWSAlgorithm.RS256, JWSAlgorithm.RS384, JWSAlgorithm.RS512 ->
                RSASSASigner(privateKey as RSAPrivateKey)

            JWSAlgorithm.ES256, JWSAlgorithm.ES384, JWSAlgorithm.ES512 ->
                ECDSASigner(privateKey as ECPrivateKey)

            else -> throw IllegalStateException("Unsupported algorithm: $algorithm")
        }

        jwt.sign(signer)
        return jwt.serialize()
    }

    /**
     * Creates a Spring [ClientHttpRequestInterceptor] that sets the
     * `Authorization: Bearer <jwt>` header on every outgoing request.
     * A fresh token is created for each request.
     */
    fun interceptor(): ClientHttpRequestInterceptor = JwtSignerInterceptor(this)

    class Builder {
        private var consumerId: String? = null
        private var privateKey: PrivateKey? = null
        private var tokenLifetime: Duration = Duration.ofSeconds(60)

        /** Sets the consumer ID used as the JWT `iss` claim. */
        fun consumerId(consumerId: String) = apply {
            require(consumerId.isNotBlank()) { "consumerId must not be blank" }
            this.consumerId = consumerId
        }

        /** Sets the private key used to sign tokens. */
        fun privateKey(privateKey: PrivateKey) = apply {
            this.privateKey = privateKey
        }

        /** Sets the token lifetime (default: 60 seconds). */
        fun tokenLifetime(lifetime: Duration) = apply {
            require(!lifetime.isNegative && !lifetime.isZero) { "tokenLifetime must be positive" }
            this.tokenLifetime = lifetime
        }

        fun build(): JwtSigner {
            val id = requireNotNull(consumerId) { "consumerId is required" }
            val key = requireNotNull(privateKey) { "privateKey is required" }
            val algo = detectAlgorithm(key)

            return JwtSigner(
                consumerId = id,
                privateKey = key,
                algorithm = algo,
                tokenLifetime = tokenLifetime,
            )
        }
    }

    private class JwtSignerInterceptor(
        private val jwtSigner: JwtSigner,
    ) : ClientHttpRequestInterceptor {
        override fun intercept(
            request: HttpRequest,
            body: ByteArray,
            execution: ClientHttpRequestExecution,
        ): ClientHttpResponse {
            request.headers.setBearerAuth(jwtSigner.createToken())
            return execution.execute(request, body)
        }
    }
}

File: contracts/api/clients/nodejs-fetch/src/auth/jwtSigner.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import { createPrivateKey, randomUUID, sign, type KeyObject } from 'node:crypto'
import { readFileSync } from 'node:fs'

/** The JOSE algorithms the signer supports, decided by the key type. */
export type JwtAlgorithm = 'RS256' | 'ES256'

const DEFAULT_TOKEN_LIFETIME_SECONDS = 60

/**
 * Mints the short-lived self-signed JWTs Epistola accepts for consumer authentication.
 *
 * Each token carries:
 * - `iss` — the consumer ID
 * - `iat` — issued-at
 * - `exp` — `iat` plus the token lifetime
 * - `jti` — a fresh UUID per token, for replay protection
 *
 * Signed with `node:crypto` alone — RS256 for RSA keys, ES256 for EC P-256 — so this needs no JOSE
 * dependency. ES256 signatures are emitted as the raw `R || S` pair JOSE requires (`ieee-p1363`),
 * not the DER sequence Node produces by default, which every compliant verifier rejects.
 *
 * The client calls {@link createToken} per request so an expired token is never sent.
 *
 * ```ts
 * const signer = JwtSigner.builder()
 *   .consumerId('invoice-service')
 *   .privateKey(JwtSigner.loadPrivateKey('private.pem'))
 *   .build()
 * ```
 */
export class JwtSigner {
  private constructor(
    private readonly consumerId: string,
    private readonly privateKey: KeyObject,
    /** The JOSE algorithm this signer uses, decided by the key type. */
    readonly algorithm: JwtAlgorithm,
    private readonly tokenLifetimeSeconds: number,
  ) {}

  /** Creates a new {@link JwtSignerBuilder}. */
  static builder(): JwtSignerBuilder {
    return new JwtSignerBuilder()
  }

  /** Loads a private key from a PEM file (RSA or EC P-256, PKCS#8 `BEGIN PRIVATE KEY`). */
  static loadPrivateKey(path: string): KeyObject {
    return JwtSigner.parsePrivateKeyPem(readFileSync(path))
  }

  /** Parses a PEM-encoded private key (RSA or EC P-256, PKCS#8 `BEGIN PRIVATE KEY`). */
  static parsePrivateKeyPem(pem: string | Buffer): KeyObject {
    try {
      return createPrivateKey(pem)
    } catch (cause) {
      throw new TypeError('Failed to parse private key. Supported formats: RSA, EC (P-256) in PKCS#8 PEM format.', { cause })
    }
  }

  /** Creates a freshly signed JWT with a new `iat`, `exp` and `jti`. */
  createToken(): string {
    const nowSeconds = Math.floor(Date.now() / 1000)
    const header = { alg: this.algorithm, typ: 'JWT' }
    const claims = {
      iss: this.consumerId,
      iat: nowSeconds,
      exp: nowSeconds + this.tokenLifetimeSeconds,
      jti: randomUUID(),
    }
    const signingInput = `${base64url(JSON.stringify(header))}.${base64url(JSON.stringify(claims))}`
    return `${signingInput}.${this.sign(signingInput).toString('base64url')}`
  }

  private sign(signingInput: string): Buffer {
    const data = Buffer.from(signingInput, 'ascii')
    return this.algorithm === 'ES256'
      ? sign('sha256', data, { key: this.privateKey, dsaEncoding: 'ieee-p1363' })
      : sign('sha256', data, this.privateKey)
  }

  /** @internal */
  static assemble(consumerId: string, privateKey: KeyObject, tokenLifetimeSeconds: number): JwtSigner {
    return new JwtSigner(consumerId, privateKey, detectAlgorithm(privateKey), tokenLifetimeSeconds)
  }
}

/** Fluent builder for {@link JwtSigner}. */
export class JwtSignerBuilder {
  private consumerIdValue: string | undefined
  private privateKeyValue: KeyObject | undefined
  private tokenLifetimeSecondsValue = DEFAULT_TOKEN_LIFETIME_SECONDS

  /** The consumer ID used as the JWT `iss` claim. */
  consumerId(consumerId: string): this {
    if (consumerId.trim() === '') {
      throw new RangeError('consumerId must not be blank')
    }
    this.consumerIdValue = consumerId
    return this
  }

  /**
   * The private key used to sign tokens, from {@link JwtSigner.loadPrivateKey} or
   * {@link JwtSigner.parsePrivateKeyPem}.
   */
  privateKey(privateKey: KeyObject): this {
    this.privateKeyValue = privateKey
    return this
  }

  /** Token lifetime in seconds (default: 60). */
  tokenLifetimeSeconds(seconds: number): this {
    if (!(seconds > 0)) {
      throw new RangeError('tokenLifetimeSeconds must be positive')
    }
    this.tokenLifetimeSecondsValue = seconds
    return this
  }

  /** Builds the immutable {@link JwtSigner}. */
  build(): JwtSigner {
    if (this.consumerIdValue === undefined) {
      throw new RangeError('consumerId is required')
    }
    if (this.privateKeyValue === undefined) {
      throw new RangeError('privateKey is required')
    }
    return JwtSigner.assemble(this.consumerIdValue, this.privateKeyValue, this.tokenLifetimeSecondsValue)
  }
}

function detectAlgorithm(key: KeyObject): JwtAlgorithm {
  if (key.type !== 'private') {
    throw new TypeError(`A private key is required, got a ${key.type} key`)
  }
  switch (key.asymmetricKeyType) {
    case 'rsa':
      return 'RS256'
    case 'ec': {
      const curve = key.asymmetricKeyDetails?.namedCurve
      if (curve !== 'prime256v1') {
        throw new TypeError(`Unsupported EC curve: ${curve ?? 'unknown'}. Supported: P-256 (prime256v1)`)
      }
      return 'ES256'
    }
    default:
      throw new TypeError(`Unsupported key type: ${key.asymmetricKeyType ?? 'unknown'}. Supported: RSA (2048+), EC (P-256)`)
  }
}

function base64url(value: string): string {
  return Buffer.from(value, 'utf8').toString('base64url')
}
